crypto/fipsmodule/ec/ec.c

```diff
@@ -391,46 +407,103 @@ EC_GROUP *EC_GROUP_new_by_curve_name(int nid) {
+      break;
+    case NID_secp256k1:
+      ret = (EC_GROUP *)OPENSSL_memdup(EC_group_secp256k1(), sizeof(EC_GROUP));
+      break;
+    default:
+      OPENSSL_PUT_ERROR(EC, EC_R_UNKNOWN_GROUP);
+      return NULL;
+  }
+  if (ret == NULL) {
+    return NULL;
+  }
+  ret->mutable_ec_group = 1;
+  return ret;
+}
+
 void EC_GROUP_free(EC_GROUP *group) {
-  if (group == NULL ||
-      // Built-in curves are static.
-      group->curve_name != NID_undef ||
-      !CRYPTO_refcount_dec_and_test_zero(&group->references)) {
+  if (group == NULL) {
     return;
   }
 
+  if (!group->mutable_ec_group) {
+    if (group->curve_name != NID_undef ||
+        !CRYPTO_refcount_dec_and_test_zero(&group->references)) {
+      // Built-in curves are static.
+      return;
+    }
+  }
+
   bn_mont_ctx_cleanup(&group->order);
   bn_mont_ctx_cleanup(&group->field);
   OPENSSL_free(group);
 }
 
 EC_GROUP *EC_GROUP_dup(const EC_GROUP *a) {
-  if (a == NULL ||
-      // Built-in curves are static.
-      a->curve_name != NID_undef) {
+  if (a == NULL) {
     return (EC_GROUP *)a;
   }
 
-  // Groups are logically immutable (but for |EC_GROUP_set_generator| which must
-  // be called early on), so we simply take a reference.
-  EC_GROUP *group = (EC_GROUP *)a;
-  CRYPTO_refcount_inc(&group->references);
-  return group;
-}
+  if (!a->mutable_ec_group) {
+    if (a->curve_name != NID_undef) {
+      // Built-in curves are static.
+      return (EC_GROUP *)a;
+    }
 
-int EC_GROUP_cmp(const EC_GROUP *a, const EC_GROUP *b, BN_CTX *ignored) {
-  // Note this function returns 0 if equal and non-zero otherwise.
-  if (a == b) {
-    return 0;
+    // Groups are logically immutable (but for |EC_GROUP_set_generator| which
+    // must be called early on), so we simply take a reference.
+    EC_GROUP *group = (EC_GROUP *)a;
+    CRYPTO_refcount_inc(&group->references);
+    return group;
   }
-  if (a->curve_name != b->curve_name) {
-    return 1;
+
+  // Directly duplicate the |EC_GROUP| if it was dynamically allocated. We do a
+  // shallow copy first, then deep copy the elements that have nested pointer
+  // redirections.
+  EC_GROUP *ret = OPENSSL_memdup(a, sizeof(EC_GROUP));
+  if (ret == NULL) {
+    return NULL;
   }
-  if (a->curve_name != NID_undef) {
-    // Built-in curves may be compared by curve name alone.
-    return 0;
+  ret->generator.group = ret;
+  bn_mont_ctx_init(&ret->field);
+  bn_mont_ctx_init(&ret->order);
+  if (!BN_MONT_CTX_copy(&ret->field, &a->field) ||
+      !BN_MONT_CTX_copy(&ret->order, &a->order)) {
+    EC_GROUP_free(ret);
+    ret = NULL;
   }
+  return ret;
+}
 
+int EC_GROUP_cmp(const EC_GROUP *a, const EC_GROUP *b, BN_CTX *ignored) {
+  if (!a->mutable_ec_group && !b->mutable_ec_group) {
+    // Note this function returns 0 if equal and non-zero otherwise.
+    if (a == b) {
+      return 0;
+    }
+    if (a->curve_name != b->curve_name) {
+      return 1;
+    }
+    if (a->curve_name != NID_undef) {
```

Comment:
nit: so, the idea here is that if both static/built-in curves have `curve_name == NID_undef`, we want to fall through to comparing individual parameters? this a bit confusing, you may want to either add a comment explaining the intent or re-write L449-504 to clarify.

Reply:
I see what the confusion is. I think it'd make sense to move this comment higher and re-clarify the intent behind returning 0 here. As you were saying, the fall through to comparing custom curve contents is what we're trying to convey.